File: runtime/lang/jp/buff.lua
buff['1'] = {
  name = "聖なる盾",
  message_0 = "は光り輝いた。",
  message_1 = "",
  description = function(self, power)
    return "PVを" .. self._effect(power) .. "上昇/耐恐怖"
  end,
}
buff['2'] = {
  name = "沈黙の霧",
  message_0 = "はぼやけた霧に覆われた。",
  message_1 = "",
  description = function(self, power)
    return "魔法の使用を禁止"
  end,
}
buff['3'] = {
  name = "リジェネレーション",
  message_0 = "の代謝が活性化した。",
  message_1 = "",
  description = function(self, power)
    return "自然回復強化"
  end,
}
buff['4'] = {
  name = "元素保護",
  message_0 = "は元素への耐性を得た。",
  message_1 = "",
  description = function(self, power)
    return "炎冷気電撃耐性の獲得"
  end,
}
buff['5'] = {
  name = "加速",
  message_0 = "は機敏になった。",
  message_1 = "",
  description = function(self, power)
    return self._effect(power) .. "の加速"
  end,
}
buff['6'] = {
  name = "鈍足",
  message_0 = "は鈍重になった。",
  message_1 = "",
  description = function(self, power)
    return self._effect(power) .. "の鈍足"
  end,
}
buff['7'] = {
  name = "英雄",
  message_0 = "の士気が向上した。",
  message_1 = "",
  description = function(self, power)
    return "筋力・器用を" .. self._effect(power) .. "上昇/耐恐怖/耐混乱"
  end,
}
buff['8'] = {
  name = "脆弱の霧",
  message_0 = "は脆くなった。",
  message_1 = "",
  description = function(self, power)
    return "DVとPVを半減"
  end,
}
buff['9'] = {
  name = "元素の傷跡",
  message_0 = "は元素への耐性を失った。",
  message_1 = "",
  description = function(self, power)
    return "炎冷気電撃耐性の減少"
  end,
}
buff['10'] = {
  name = "ホーリーヴェイル",
  message_0 = "は聖なる衣に保護された。",
  message_1 = "",
  description = function(self, power)
    return "ﾊﾟﾜｰ" .. self._effect(power) .. "の呪い(hex)への抵抗"
  end,
}
buff['11'] = {
  name = "ナイトメア",
  message_0 = "は悪夢に襲われた。",
  message_1 = "",
  description = function(self, power)
    return "神経幻惑耐性の減少"
  end,
}
buff['12'] = {
  name = "知者の加護",
  message_0 = "の思考は冴え渡った。",
  message_1 = "",
  description = function(self, power)
    a, b = self._effect(power)
    return "習得・魔力を" .. a .. "上昇/読書を" .. b .. "上昇"
  end,
}
buff['13'] = {
  name = "天罰",
  message_0 = "は雷に打たれた！",
  message_1 = "",
  description = function(self, power)
    a, b = self._effect(power)
    return a .. "の鈍足/PVを" .. b .. "%減少"
  end,
}
buff['14'] = {
  name = "ルルウィの憑依",
  message_0 = "にルルウィが乗り移った。",
  message_1 = "",
  description = function(self, power)
    return self._effect(power) .. "の加速"
  end,
}
buff['15'] = {
  name = "インコグニート",
  message_0 = "は別人になりすました。",
  message_1 = "",
  description = function(self, power)
    return "変装"
  end,
}
buff['16'] = {
  name = "死の宣告",
  message_0 = "は死の宣告を受けた！",
  message_1 = "",
  description = function(self, power)
    return "呪いが完了したときに確実なる死"
  end,
}
buff['17'] = {
  name = "ブースト",
  message_0 = "はブーストした！",
  message_1 = "",
  description = function(self, power)
    return self._effect(power) .. "の加速と能力のアップ"
  end,
}
buff['18'] = {
  name = "契約",
  message_0 = "は死神と契約した。",
  message_1 = "",
  description = function(self, power)
    return "致命傷を負ったとき" .. self._effect(power) .. "%の確率でダメージ分回復。"
  end,
}
buff['19'] = {
  name = "幸運",
  message_0 = "に幸運な日が訪れた！",
  message_1 = "",
  description = function(self, power)
    return self._effect(power) .. "の幸運の上昇"
  end,
}
buff['20'] = {
  name = "筋力の成長",
  message_0 = "",
  message_1 = "",
  description = function(self, power)
    return ability['10'].name .. "の成長率を" .. self._effect(power) .. "%上昇"
  end,
}
buff['21'] = {
  name = "耐久の成長",
  message_0 = "",
  message_1 = "",
  description = function(self, power)
    return ability['11'].name .. "の成長率を" .. self._effect(power) .. "%上昇"
  end,
}
buff['22'] = {
  name = "器用の成長",
  message_0 = "",
  message_1 = "",
  description = function(self, power)
    return ability['12'].name .. "の成長率を" .. self._effect(power) .. "%上昇"
  end,
}
buff['23'] = {
  name = "感覚の成長",
  message_0 = "",
  message_1 = "",
  description = function(self, power)
    return ability['13'].name .. "の成長率を" .. self._effect(power) .. "%上昇"
  end,
}
buff['24'] = {
  name = "習得の成長",
  message_0 = "",
  message_1 = "",
  description = function(self, power)
    return ability['14'].name .. "の成長率を" .. self._effect(power) .. "%上昇"
  end,
}
buff['25'] = {
  name = "意思の成長",
  message_0 = "",
  message_1 = "",
  description = function(self, power)
    return ability['15'].name .. "の成長率を" .. self._effect(power) .. "%上昇"
  end,
}
buff['26'] = {
  name = "魔力の成長",
  message_0 = "",
  message_1 = "",
  description = function(self, power)
    return ability['16'].name .. "の成長率を" .. self._effect(power) .. "%上昇"
  end,
}
buff['27'] = {
  name = "魅力の成長",
  message_0 = "",
  message_1 = "",
  description = function(self, power)
    return ability['17'].name .. "の成長率を" .. self._effect(power) .. "%上昇"
  end,
}
buff['28'] = {
  name = "速度の成長",
  message_0 = "",
  message_1 = "",
  description = function(self, power)
    return ability['18'].name .. "の成長率を" .. self._effect(power) .. "%上昇"
  end,
}
buff['29'] = {
  name = "運勢の成長",
  message_0 = "",
  message_1 = "",
  description = function(self, power)
    return ability['19'].name .. "の成長率を" .. self._effect(power) .. "%上昇"
  end,
}



-- vim: et sw=2 sts=2

File: src/buff.cpp
#include "buff.hpp"
#include "ability.hpp"
#include "character.hpp"
#include "draw.hpp"
#include "elona.hpp"
#include "fov.hpp"
#include "i18n.hpp"
#include "lua_env/interface.hpp"
#include "message.hpp"
#include "random.hpp"
#include "variables.hpp"



using namespace elona;


namespace
{



constexpr int buff_find_slot_no_effect = -1;

int buff_find_slot(const Character& cc, int id, int turns)
{
    for (size_t i = 0; i < cc.buffs.size(); ++i)
    {
        if (cc.buffs[i].id == id)
        {
            if (cc.buffs[i].turns < turns)
            {
                return static_cast<int>(i);
            }
            else
            {
                return buff_find_slot_no_effect;
            }
        }
        if (cc.buffs[i].id == 0)
        {
            return i;
        }
    }

    return rnd(static_cast<int>(cc.buffs.size()));
}

void buff_on_removal(Character& chara, int id)
{
    auto buff = the_buff_db[id];
    assert(buff);

    auto& on_removal = buff->on_removal;
    if (on_removal)
    {
        auto self = buff->self;
        auto args = lua::create_table("chara", lua::handle(chara));
        on_removal->call(self, args);
    }
}



} // namespace



namespace elona
{


int calc_buff_duration(int id, int power)
{
    auto buff = the_buff_db[id];
    assert(buff);

    auto& duration = buff->duration;
    return duration.call_with_result(0, power);
}


std::string get_buff_description(int id, int power)
{
    auto buff = the_buff_db[id];
    assert(buff);

    auto& self = buff->self;
    auto& description = buff->description;
    return description.call_with_result("<error>", self, power);
}



void apply_buff(int cc, int id, int power)
{
    auto buff = the_buff_db[id];
    assert(buff);

    auto& self = buff->self;
    auto& on_refresh = buff->on_refresh;
    auto args =
        lua::create_table("power", power, "chara", lua::handle(cdata[cc]));
    on_refresh.call(self, args);
}


bool buff_has(const Character& cc, const std::string& id)
{
    auto buff_def = the_buff_db[id];
    assert(buff_def);

    int legacy_id = buff_def->id;
    return std::any_of(
        std::begin(cc.buffs), std::end(cc.buffs), [&](const auto& buff) {
            return buff.id == legacy_id;
        });
}



optional_ref<const Buff> buff_find(const Character& cc, const std::string& id)
{
    auto buff_def = the_buff_db[id];
    assert(buff_def);

    int legacy_id = buff_def->id;
    const auto itr = std::find_if(
        std::begin(cc.buffs), std::end(cc.buffs), [&](const auto& buff) {
            return buff.id == legacy_id;
        });
    if (itr == std::end(cc.buffs))
    {
        return none;
    }
    else
    {
        return *itr;
    }
}



void buff_add(
    Character& cc,
    const std::string& id,
    int power,
    int turns,
    optional_ref<const Character> doer)
{
    auto buff = the_buff_db[id];
    assert(buff);

    int legacy_id = buff->id;

    if (turns <= 0)
        return;

    const auto slot = buff_find_slot(cc, legacy_id, turns);
    if (slot == buff_find_slot_no_effect)
    {
        if (is_in_fov(cc))
        {
            txt(i18n::s.get("core.locale.magic.buff.no_effect"));
            return;
        }
    }

    if (buff->type == BuffType::hex)
    {
        bool resists{};
        if (sdata(60, cc.index) / 2 > rnd(power * 2 + 100))
        {
            resists = true;
        }
        if (power * 3 < sdata(60, cc.index))
        {
            resists = true;
        }
        if (power / 3 > sdata(60, cc.index))
        {
            resists = false;
        }
        if (cc.quality > Quality::great)
        {
            if (rnd(4))
            {
                resists = true;
            }
            else
            {
                turns = turns / 5 + 1;
            }
        }
        if (cc.quality >= Quality::miracle && id == "core.death_word")
        {
            resists = true;
        }
        if (const auto& holy_veil = buff_find(cc, "core.holy_veil"))
        {
            if (holy_veil->power + 50 > power * 5 / 2 ||
                rnd(holy_veil->power + 50) > rnd(power + 1))
            {
                txt(i18n::s.get("core.locale.magic.buff.holy_veil_repels"));
                return;
            }
        }
        if (resists)
        {
            if (is_in_fov(cc))
            {
                txt(i18n::s.get("core.locale.magic.buff.resists", cc));
            }
            return;
        }
        if (doer && doer->index == 0)
        {
            hostileaction(0, cc.index);
        }
    }

    if (is_in_fov(cc))
    {
        // Messages of fodd buff are shown elsewhere.
        if (buff->type != BuffType::food)
        {
            txt(i18n::s.get_enum_property(
                "core.locale.buff", legacy_id, "apply", cc));
        }

        add_damage_popup(
            i18n::s.get_enum_property("core.locale.buff", "name", legacy_id),
            cc.index,
            {255, 255, 255});
    }

    cc.buffs[slot] = {legacy_id, power, turns};

    chara_refresh(cc.index);
}



void buff_delete(Character& cc, int slot)
{
    if (cc.index == 0)
    {
        txt(i18n::s.get(
                "core.locale.magic.buff.ends",
                i18n::s.get_enum_property(
                    "core.locale.buff", "name", cc.buffs[slot].id)),
            Message::color{ColorIndex::purple});
    }
    if (is_in_fov(cc))
    {
        add_damage_popup(
            i18n::s.get_enum_property(
                "core.locale.buff", "name", cc.buffs[slot].id),
            cc.index,
            {191, 191, 191});
    }
    buff_on_removal(cc, cc.buffs[slot].id);
    cc.buffs[slot].id = 0;
    for (int cnt = slot, cnt_end = cnt + (16 - slot - 1); cnt < cnt_end; ++cnt)
    {
        if (cc.buffs[cnt].id == 0)
        {
            if (cc.buffs[cnt + 1].id != 0)
            {
                cc.buffs[cnt] = cc.buffs[cnt + 1];
                cc.buffs[cnt + 1] = Buff{};
            }
            else
            {
                break;
            }
        }
    }

    chara_refresh(cc.index);
}



} // namespace elona
